core/apple/src/AppleCore.kt

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.NSInputStream
+import platform.Foundation.NSOutputStream
+import platform.Foundation.NSStreamStatusNotOpen
+import platform.darwin.UInt8Var
+
+/**
+ * Returns [RawSink] that writes to an output stream.
+ *
+ * Use [RawSink.buffered] to create a buffered sink from it.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.outputStreamAsSink
+ */
+public fun NSOutputStream.asSink(): RawSink = OutputStreamSink(this)
+
+private open class OutputStreamSink(
+    private val out: NSOutputStream,
+) : RawSink {
+
+    @OptIn(UnsafeNumber::class)
+    override fun write(source: Buffer, byteCount: Long) {
+        if (out.streamStatus == NSStreamStatusNotOpen) out.open()
+
+        checkOffsetAndCount(source.size, 0, byteCount)
+        var remaining = byteCount
+        while (remaining > 0) {
+            val head = source.head!!
+            val toCopy = minOf(remaining, head.limit - head.pos).toInt()
+            val bytesWritten = head.data.usePinned {
+                val bytes = it.addressOf(head.pos).reinterpret<UInt8Var>()
+                out.write(bytes, toCopy.convert()).toLong()
+            }
+
+            if (bytesWritten < 0L) throw IOException(out.streamError?.localizedDescription ?: "Unknown error")
```

Comment:
Im not sure, but may be should use just `description` and not `localizedDescription` - I don't have preferences, and don't really know, how it works on JVM in such cases

Reply:
[`NSError.localizedDescription`](https://developer.apple.com/documentation/foundation/nserror/1414418-localizeddescription?language=objc) is similar to [`Throwable.message`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-throwable/message.html) on the JVM, where [`NSObject.description`](https://developer.apple.com/documentation/objectivec/1418956-nsobject/1418746-description?language=objc) is the equivalent of [`Any.toString()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-any/to-string.html). I actually remember this being confusing myself when I was first learning Objective-C for iOS, having more experience with Java a decade ago.